### PR TITLE
Qlib simulator refinement

### PR DESCRIPTION
**Pull request by lihuoran**
## Description

- Qlib Simulator
  - Use dict-like config to create strategies & executors. Totally abandon the hardcoded version.
  - Make Qlib simulator more general.
- Executor
  - Add `is_collecting`
  - Call `inner_strategy.post_upper_level_exe_step()` in NestedExecutor
- Strategy
  - Add `post_upper_level_exe_step()` method
  - Add `SAOEStrategy`, strategies that use SAOEState as states. Make `DecomposedStrategy` a subclass of `SAOEStrategy`.
  - Move `StateMaintainer` inside `SAOEStrategy`. Now each `SAOEStrategy` can have its own state maintainer, so it can maintain its own state.
- Other
  - Remove `qlib/rl/from_neutrader`. Move all necessary components into `simulator_qlib`.
  - Re-organize file structure
    - `qlib/rl/strategy`
    - `qlib/rl/order_execution/objects`
    - `qlib/rl/order_execution/state`

## How Has This Been Tested?

- [ ] Pass the test by running: `pytest qlib/tests/test_all_pipeline.py` under upper directory of `qlib`.
- [ ] If you are adding a new feature, test on your own test scripts.

## Screenshots of Test Results (if appropriate):
1. Pipeline test:
2. Your own tests:

## Types of changes

- [ ] Fix bugs
- [ ] Add new feature
- [ ] Update documentation

### CODE REVIEW

File: qlib/rl/order_execution/state.py [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+
+from qlib.backtest import Order
+from qlib.backtest.executor import BaseExecutor
+from qlib.constant import EPS
+from qlib.rl.data.exchange_wrapper import QlibIntradayBacktestData
+from qlib.rl.data.pickle_styled import IntradayBacktestData
+from qlib.rl.order_execution.utils import dataframe_append, get_simulator_executor, price_advantage
+
+
+class SAOEStateMaintainer:
```
Comment (ultmaster) [inline qlib/rl/order_execution/state.py:20]: QlibBacktestAdapter?

File: qlib/rl/strategy/saoe.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+
+from abc import ABCMeta
```
Comment (ultmaster) [inline qlib/rl/strategy/saoe.py:4]: I suggest SAOE-related files be put into `qlib/rl/order_execution`.

File: qlib/rl/order_execution/simulator_qlib.py [hunk 3]
```diff
@@ -3,261 +3,94 @@
...
+def create_state_maintainer_recursive(
+    executor: BaseExecutor,
+    order: Order,
+    backtest_data: QlibIntradayBacktestData,
+    time_per_step: str,
+    ticks_index: pd.DatetimeIndex,
+    twap_price: float,
+    ticks_for_order: pd.DatetimeIndex,
+) -> None:
+    if isinstance(executor, SimulatorExecutor):
```
[the remark this answers is not in the document]
Comment (lihuoran) [reply qlib/rl/order_execution/simulator_qlib.py:69]: This function was deleted.

File: qlib/rl/order_execution/simulator_qlib.py [hunk 4]
```diff
@@ -272,112 +105,95 @@ class SingleAssetOrderExecutionQlib(Simulator[Order, SAOEState, float]):
         A string to describe the time granularity of each step. Current support "1min", "30min", and "1day"
     qlib_config (dict):
         Configuration used to initialize Qlib.
-    inner_executor_fn (Callable[[str, CommonInfrastructure], BaseExecutor]):
-        Function used to get the inner level executor.
-    exchange_config (ExchangeConfig):
-        Configuration used to create the Exchange instance.
+    strategy_config (dict):
```
[the remark this answers is not in the document]
Comment (lihuoran) [reply qlib/rl/order_execution/simulator_qlib.py]: Could you please review the latest version and provide some details?

File: qlib/rl/order_execution/simulator_qlib.py [hunk 5]
```diff
@@ -3,381 +3,158 @@
...
-
-        maintainer = StateMaintainer(...)  # in reset
-        maintainer.update(...)  # in step
-        # get states in get_state from maintainer
+    Parameters
+    ----------
+    qlib_config:
+        Qlib configuration.
+
+        Example:
```
Comment (ultmaster) [inline qlib/rl/order_execution/simulator_qlib.py:28]: `Example::`  Double semicolon for code-block. Blank line follows.

File: qlib/rl/order_execution/strategy.py [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+from qlib.backtest import CommonInfrastructure, Order
+from qlib.backtest.decision import BaseTradeDecision, TradeDecisionWO, TradeRange, TradeRangeByTime
+from qlib.backtest.utils import LevelInfrastructure, SAOE_DATA_KEY
+from qlib.rl.data.exchange_wrapper import QlibIntradayBacktestData
+from qlib.rl.order_execution.state import QlibBacktestAdapter, SAOEState
+from qlib.rl.order_execution.utils import get_ticks_slice
+from qlib.strategy.base import RLStrategy
+
+
+class SAOEStrategy(RLStrategy, metaclass=ABCMeta):
```
Comment (ultmaster) [inline qlib/rl/order_execution/strategy.py:18]: Why is it ABCMeta?

File: qlib/backtest/utils.py [hunk 7]
```diff
@@ -21,6 +21,9 @@
 from ..data.data import Cal
 
 
+SAOE_DATA_KEY = "saoe_data"
```
Comment (ultmaster) [inline qlib/backtest/utils.py:24]: Where is this used?

File: qlib/rl/strategy/single_order.py [hunk 8]
```diff
@@ -0,0 +1,35 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+
+from qlib.backtest import Order
+from qlib.backtest.decision import BaseTradeDecision, OrderHelper, TradeDecisionWO, TradeRange
+from qlib.strategy.base import BaseStrategy
+
+
+class SingleOrderStrategy(BaseStrategy):
```
Comment (ultmaster) [inline qlib/rl/strategy/single_order.py:9]: Add docstrings for this.

File: qlib/strategy/base.py [hunk 9]
```diff
@@ -55,10 +58,18 @@ def __init__(
...
+    def executor(self) -> BaseExecutor:
+        return self.level_infra.get("executor")
+
     @property
     def trade_calendar(self) -> TradeCalendarManager:
         return self.level_infra.get("trade_calendar")
 
+    @property
+    def ticks_per_step(self) -> int:
+        return int(pd.Timedelta(self.trade_calendar.get_freq()) / pd.Timedelta("1min"))
```
Comment (ultmaster) [inline qlib/strategy/base.py:71]: 1min is FINEST_GRANULARITY.

File: qlib/rl/order_execution/strategy.py [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+                ticks_for_order = None  # FIXME: implement this logic
+
+            backtest_data = QlibIntradayBacktestData(
+                order=order,
+                exchange=self.trade_exchange,
+                start_time=ticks_for_order[0],
+                end_time=ticks_for_order[-1],
+            )
+
+            saoe_data[order.key] = (ticks_index, ticks_for_order, backtest_data)
```
Comment (ultmaster) [inline qlib/rl/order_execution/strategy.py:67]: Would `saoe_data` get larger and larger and overwhelm the memory?

File: qlib/backtest/utils.py (hunk 7 above)
Comment (lihuoran) [reply qlib/backtest/utils.py:24]: For `SAOEStrategy`.

File: qlib/rl/order_execution/strategy.py (hunk 10 above)
Comment (lihuoran) [reply qlib/rl/order_execution/strategy.py:67]: That's a good point. I use "LRUCache" to store this data now.

File: qlib/backtest/utils.py [hunk 11]
```diff
@@ -235,7 +238,9 @@ def update(self, other: BaseInfrastructure) -> None:
 
 class CommonInfrastructure(BaseInfrastructure):
     def get_support_infra(self) -> Set[str]:
-        return {"trade_account", "trade_exchange"}
+        # SAOE_DATA_KEY is used to store SAOE (single asset order execution) information that should be shared by
```
[the remark this answers is not in the document]
Comment (lihuoran) [reply qlib/backtest/utils.py]: We changed the design here and use LRU cache to store data in the latest version.

File: qlib/backtest/executor.py [hunk 12]
```diff
@@ -124,6 +124,9 @@ def __init__(
         self.dealt_order_amount: Dict[str, float] = defaultdict(float)
         self.deal_day = None
 
+        # whether the current executor is collecting data
+        self.is_collecting = False
```
[the remark this answers is not in the document]
Comment (lihuoran) [reply qlib/backtest/executor.py:128]: Nice suggestion. Code changed.

File: qlib/strategy/base.py [hunk 13]
```diff
@@ -207,6 +218,12 @@ def get_data_cal_avail_range(self, rtype: str = "full") -> Tuple[int, int]:
         range_limit = self.outer_trade_decision.get_data_cal_range_limit(rtype=rtype)
         return max(cal_range[0], range_limit[0]), min(cal_range[1], range_limit[1])
 
+    def post_upper_level_exe_step(self) -> None:
```
[the remark this answers is not in the document]
Comment (lihuoran) [reply qlib/strategy/base.py:221]: I leave a more detailed explanation in the code with a TODO. Maybe we could polish the design here in the future?

File: qlib/constant.py [hunk 14]
```diff
@@ -11,3 +16,7 @@
 
 # Infinity in integer
 INF = 10**18
+FINEST_GRANULARITY = "1min"
+COARSEST_GRANULARITY = "1day"
+ONE_SEC = pd.Timedelta("1s")  # use 1 second to exclude the right interval point
```
[the remark this answers is not in the document]
Comment (lihuoran) [reply qlib/constant.py:21]: Do you have any suggestions for the naming here?

File: qlib/rl/utils/cache.py [hunk 15]
```diff
@@ -0,0 +1,24 @@
+from __future__ import annotations
```
[the remark this answers is not in the document]
Comment (lihuoran) [reply qlib/rl/utils/cache.py:1]: Use `cachetools` here.

File: qlib/strategy/base.py [hunk 16]
```diff
@@ -55,10 +60,18 @@ def __init__(
...
+    @property
+    def executor(self) -> BaseExecutor:
+        return self.level_infra.get("executor")
+
     @property
     def trade_calendar(self) -> TradeCalendarManager:
         return self.level_infra.get("trade_calendar")
 
+    @property
+    def ticks_per_step(self) -> int:
```
[the remark this answers is not in the document]
Comment (lihuoran) [reply qlib/strategy/base.py:72]: What's the difference between trading time and calendar time? What I am trying to say here is actually the "number of minutes per step".

File: qlib/rl/order_execution/strategy.py [hunk 17]
```diff
@@ -0,0 +1,125 @@
...
+        # level strategy through this way.
+        exec_vol = yield self
+
+        oh = self.trade_exchange.get_order_helper()
+        order = oh.create(self._order.stock_id, exec_vol, self._order.direction)
+        self._update_last_step_range(self.get_data_cal_avail_range(rtype="step"))
+
+        return TradeDecisionWO([order], self)
+
+    def alter_outer_trade_decision(self, outer_trade_decision: BaseTradeDecision) -> BaseTradeDecision:
```
[the remark this answers is not in the document]
Comment (lihuoran) [reply qlib/rl/order_execution/strategy.py:115]: I implemented the default behavior in the base method and remove all unnecessary overwritten methods. The error presented in the figure was also fixed.

File: qlib/backtest/decision.py [hunk 18]
```diff
@@ -135,6 +135,11 @@ def parse_dir(direction: Union[str, int, np.integer, OrderDir, np.ndarray]) -> U
         else:
             raise NotImplementedError(f"This type of input is not supported")
 
+    @property
+    def key(self) -> tuple:
+        """A hashable & unique key to identify this order. Usually used as the key in a dict."""
+        return self.stock_id, self.start_time.replace(hour=0, minute=0, second=0), self.direction
```
[the remark this answers is not in the document]
Comment (lihuoran) [reply qlib/backtest/decision.py:141]: I added a new property called `key_by_day` which is identical to the current version of `key`. As stated by the name, `key_by_day` returns the key under granularity in days, and the new version of `key` returns the tuple with `start_time` and `end_time`. 

These two properties shall be used in different scenarios. For example, when loading intraday data, we should use `key_by_day`. However, when maintaining order-related information in the strategy, we should use `key`.

File: qlib/constant.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (you-n-g) [reply qlib/constant.py:21]: @lihuoran  
I don't have any preferences about the `FINEST_GRANULARITY` and  `COARSEST_GRANULARITY` variables.  Both versions are OK to me.

My preference is for `ONE_SEC`. My preferred name is `EPSILON_TIME`

[the remark this answers is not in the document]
Comment (you-n-g) [reply qlib/constant.py:21]: `EPS_T` is also a good name

File: qlib/strategy/base.py [hunk 19]
```diff
@@ -207,6 +218,12 @@ def get_data_cal_avail_range(self, rtype: str = "full") -> Tuple[int, int]:
         range_limit = self.outer_trade_decision.get_data_cal_range_limit(rtype=rtype)
         return max(cal_range[0], range_limit[0]), min(cal_range[1], range_limit[1])
 
+    def post_upper_level_exe_step(self) -> None:
+        """
+        A hook for doing sth after the upper level executor finished its execution (for example, finalize
+        the metrics collection).
+        """
+
     def post_exe_step(self, execute_result: list) -> None:
```
[the remark this answers is not in the document]
Comment (you-n-g) [reply qlib/strategy/base.py:225]: Don't forget this

**Review by you-n-g — APPROVED** [review]

File: qlib/rl/integration/feature.py [hunk 20]
```diff
@@ -0,0 +1,163 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+
+"""
```
Comment (ultmaster) [inline qlib/rl/integration/feature.py:4]: Suggest moving this file into `order_execution/integration.py`.

Okay to keep it for now.

File: qlib/rl/order_execution/state.py [hunk 21]
```diff
@@ -0,0 +1,331 @@
...
+    if ret[-1] > end:
+        ret.pop()
+    if ret[-1] == end and not include_end:
+        ret.pop()
+    return pd.DatetimeIndex(ret)
+
+
+class SAOEStateAdapter:
+    """
+    Maintain states of the environment.
```
Comment (ultmaster) [inline qlib/rl/order_execution/state.py:40]: Could you add some explanations of "what" should be adapted and "how".

File: qlib/rl/strategy/single_order.py [hunk 22]
```diff
@@ -0,0 +1,31 @@
...
+    """Strategy used to generate a trade decision with exactly one order."""
+
+    def __init__(
+        self,
+        order: Order,
+        trade_range: TradeRange,
```
Comment (ultmaster) [inline qlib/rl/strategy/single_order.py:15]: I think `trade_range` can be None?

**Review by ultmaster — APPROVED** [review]

**Pull request merged**